docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1265 @@
+
+Which parameters could be considered?
+* Number of available machines of a type as requested as part of the shoot spec. Sufficient capacity should be available to also allow rolling updates which will also be governed by `maxSurge` configuration at the worker pool level.
+* Node CIDR range must grant enough space to schedule additional replicas that the HA feature requires.
+* Number of volumes that will be required to host a multi-node/multi-zone etcd cluster will increase by `(n-1)` where `n` is the total number of members in the etcd cluster.
+
+The above list is not an exhaustive list and is just indicative that the currently set limit of 250 will have to be revisited.
+
+## Scheduling control plane components
+
+### Zone pinning
+
+HA shoot clusters with failure tolerance of `node` as well as non-HA shoot clusters can be scheduled on `single-zonal` and `multi-zonal` seeds alike. On a `multi-zonal` seed it's desireable to place components of the same control plane in **one zone only** to reduce cost and latency effects due to cross network traffic.
+Thus, it's essential to add [Pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) rules to control plane component with multiple replicas:
+
+```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchLabels:
+            gardener.cloud/shoot-namespace: <shoot-namespace>
+            <labels>
+        topologyKey: "topology.kubernetes.io/zone"
+```
+
+A special challenge is to select the entire set of control plane pods belonging to a single control plane. Today, Gardener and extensions don't put a common label to the affected pods. We propose to introduce a new label `gardener.cloud/shoot-namespace: <shoot-namespace>`. A mutating webhoook in the __Gardener Resource Manager__ should apply this label and the affinity rules to every pod in the control plane. This label and the pod affinity rule will ensure that all the pods in the control plane are pinned to a specific zone for HA shoot cluster having failure tolerance of `node`.
+
+### Single-Zone
+
+Furthermore, for HA shoot cluster with failure tolerance of `node` the following anti-affinity rule guarantees that each pod of a control plane component is scheduled onto a different node. For the `Kube-Apiserver` this can mean that a seed cluster has to have even more than three nodes (see [Recommended-number-of-nodes-and-zones](#recommended-number-of-nodes-and-zones)) by the time of scheduling if `HPA` scales the deployment beyond three replicas. This should be common for most of the seeds running in production. In addition, we think that fulfilling the node spread is more important for HA shoots than supporting seeds with the absolute minimum worker count in that edge case.
```

Comment:
Regarding kube-apiserver and scale up (`Single-Zone (#replicas > #nodes)`):
I think we should follow a similar pattern as described in `Multi-Zone (#replicas > #zones)` instead of specifying hard anti-affinity, i.e. require that replicas are spread across at least 3 nodes, but don't require 5 replicas to be spread across 5 nodes.
Otherwise, hard anti-affinity will again limit cluster-autoscaler's options for scale-down and thereby decrease seed cluster utilization further.

Reply:
For a `single-zone` the idea was that there should not be more than one instance per node which ensures that at least a node failure is tolerated. If we use TSC for a single-zone then it will look like the following:
```yaml
spec:
  topologySpreadConstraints:
  - maxSkew: 1
     topologyKey: kubernetes.io/hostname
     whenUnsatisfiable: DoNotSchedule
     labelSelector:
      matchLabels:
        <labels>
```
So if the number of replicas are let's say 4, then you could have more than 1 replica on a single node. Is that acceptable and desirable? From the perspective of tolerance to node failure, there will still be tolerance as (maybe not as good as one-replica-per-node), however from the perspective of CA there will be better utilisation of resources.

Reply:
That sounds good 👍

Reply:
addressed. Please check now.